File: Sliding Window/leetcode_220.cpp

// Problem link ===>> https://leetcode.com/problems/contains-duplicate-iii/description/?envType=problem-list-v2&envId=21r1z5od

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
#define ll long long
#define ld long double
#define endl '\n'
const int MOD = 1e9 + 7;
const double EPS = 1e-9;
const double PI = acos(-1.0);
template <typename T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
template <typename T>
using ordered_set_mul = tree<T, null_type, less_equal<T>, rb_tree_tag, tree_order_statistics_node_update>;

class Solution
{
public:
    bool containsNearbyAlmostDuplicate(vector<int> &nums, int k, int t)
    {
        int n = nums.size();
        if(n==0 || k<0 || t<0){
            return false;
        }

        unordered_map<int, int> buckets;
        for (int i = 0; i < n; i++){
            int bucket = nums[i] / ((long)t + 1);

            if(nums[i]<0)
                --bucket;
            
            if(buckets.find(bucket) != buckets.end()){
                return true;
            } else {
                buckets[bucket] = nums[i];

                if(buckets.find(bucket - 1) != buckets.end() && abs((long)nums[i] - buckets[bucket - 1]) <= t){
                    return true;
                }
                if(buckets.find(bucket + 1) != buckets.end() && abs((long)nums[i] - buckets[bucket + 1]) <= t){
                    return true;
                }

                if (buckets.size() > k)
                {
                    int key_to_remove = nums[i - k] / ((long)t + 1);

                    if (nums[i - k] < 0)
                        --key_to_remove;

                    buckets.erase(key_to_remove);
                }
            }
        }
        return false;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);

    ll t;
    cin >> t;
    while (t--)
    {
        ll n;
        cin >> n;
    }
    return 0;
}

File: Sliding Window/leetcode_220.py

# Problem link ===>> https://leetcode.com/problems/contains-duplicate-iii/description/?envType=problem-list-v2&envId=21r1z5od
# submission link == =>> https://leetcode.com/problems/contains-duplicate-iii/submissions/1700254598/?envType=problem-list-v2&envId=21r1z5od

import collections
import sys
import math
import bisect
import heapq
from collections import defaultdict, deque, OrderedDict, Counter
from functools import lru_cache, reduce
from itertools import permutations, combinations, accumulate
from typing import List, Tuple, Dict, Set, Optional, Union, Any

MOD = 10**9 + 7
INF = float('inf')
input = sys.stdin.read
print = sys.stdout.write


class Solution:
    def containsNearbyAlmostDuplicate(self, nums: list[int], k: int, t: int) -> bool:
        n = len(nums)
        if n == 0 or k < 0 or t < 0:
            return False
        
        bucket_size = t + 1

        buckets = {}

        for i in range(n):
            num = nums[i]

            bucket_idx = num // bucket_size
            if num < 0 and num % bucket_size != 0:
                bucket_idx = int(num / bucket_size)
                pass
            
            bucket_idx = int(num / bucket_size)
            if num < 0:
                bucket_idx -= 1

            if bucket_idx in buckets:
                return True

            if (bucket_idx - 1) in buckets and abs(num - buckets[bucket_idx - 1]) <= t:
                return True
            
            if (bucket_idx + 1) in buckets and abs(num - buckets[bucket_idx + 1]) <= t:
                return True

            buckets[bucket_idx] = num

            if i >= k:
                old_num = nums[i - k]
                old_bucket_idx = int(old_num / bucket_size)
                
                if old_num < 0:
                    old_bucket_idx -= 1
                    
                if old_bucket_idx in buckets:
                    del buckets[old_bucket_idx]

        return False

def main():
    t = int(input())
    for _ in range(t):
        n = int(input())

if __name__ == "__main__":
    solution = Solution()
